### ISSUE TITLE
`"editor.suggestSelection": "recentlyUsed"` doesn't pick the most recently-used option

### ISSUE THREAD

**Issue by varkor** [microsoft/vscode#78571]
- VSCode Version: 1.36.1
- OS Version: macOS 10.12.6

Consider the file:
```
game
gamma
```

1. Make sure `"editor.suggestSelection": "recentlyUsed"`.
1. On a new line, type `g` then select `gamma` in the autocomplete list.
1. Press space and type `g` again. `game` will be selected, despite the setting being `recentlyUsed`.

I've also encountered an issue where an option is selected with `recentlyUsed`, then I type the next character of the suggestion, and it changes to a different suggestion, despite still matching the current suggestion. As I understand it, this should not happen with `recentlyUsed`. However, I was unable to come up with a test case for this example.

**Comment by jrieken** [comment]
What you want is the `recentlyUsedByPrefix`

**Comment by varkor** [comment]
`recentlyUsedByPrefix` does not provide the behaviour I want, because it requires the entire prefix to be typed. E.g. if I type `gam` and tab-complete to `gamma`, then typing `ga` will still suggest `game` first.

If `recentlyUsed` is acting as intended, then at the very least its description is misleading. ("Recently used" would usually imply "most recently used", but it seems "recently used" is "an arbitrary order of words that you typed recently.) What I'm looking for is a "most recently used" option. Is there a reason that is not how `recentlyUsed` works?

**Comment by jrieken** [comment]
> then at the very least its description is misleading. 

Well, I am open for suggestions. Below is the doc, it tries to say that `recentlyUsed` only works when no prefix is present ("...unless further typing...")

> Select recent suggestions unless further typing selects one, e.g. `console.| -> console.log` because `log` has been completed recently.

**Comment by varkor** [comment]
"unless further typing selects one" implies to me that the most recent choice is no longer valid in light of the prefix. Why does "g" select "game" rather than "gamma", when "gamma" is the most recently chosen suggestion and is matched by "g"?

**Comment by jrieken** [comment]
> Why does "g" select "game" rather than "gamma", when "gamma" is the most recently chosen

Because "g" means that typing has selected one, e.g that what's meant by "unless further typing..."

**Comment by varkor** [comment]
So, to check I understand, "selection" occurs after any typing, and after that the suggestion is completely ignored. I would argue that it's more intuitive for "selection" to only occur after typing a non-prefix of the suggestion, but maybe others would disagree.

It seems like what I really want, then, is a way to sort the suggestion list by recency, rather than alphabetically, so "gamma" appears before "game" if it's the most recent choice, and I type any prefix of it. Is there a way to customise this? I tried looking around and couldn't find anything. Thanks!

**Comment by jrieken** [comment]
Right, the feature is "select by recency", not "sort by recency" . When typing selects one (which I agree is a weird formulation, please help) then there is a conflict between history and the current prefix. Tho, while I am typing this we could say "when typing doesn't yield a clear winner than the recency selection remains", e.g. `g` scores equal against `game` and `gamma` so don't select the first by language sort order but select the most recent one. That might help a little

**Comment by varkor** [comment]
> "when typing doesn't yield a clear winner than the recency selection remains", e.g. g scores equal against game and gamma so don't select the first by language sort order but select the most recent one. That might help a little

I think this would address my use case, actually, as the example in my first post is indicative of the problem I have: resolving ties by recency rather than the default sort order would be a definite improvement in my opinion. If that'd be possible, that'd be great!

**Comment by jrieken** [comment]
Let's see what's possible...

**Comment by jrieken** [comment]
Ok, I will push a change to make this better be fyi that there has been https://github.com/microsoft/vscode/issues/43429. Because of that, the `recentlyUsed`-strategy doesn't select anything when there is leading whitespace. Again, we might find a better solution here but I am careful given the emotions.

This is how it works with my changes

![Aug-09-2019 14-50-27](https://user-images.githubusercontent.com/1794099/62780041-14a95780-bab5-11e9-93ff-55909eb3bd0b.gif)

The first time I select `foo.game` which is then remembered. Next time, with `foo.` the `game` suggestion is selected and it stays selected while I type. The selection changes once the `game` suggestion isn't the group of best matches anymore.

**Comment by varkor** [comment]
@jrieken: just to clarify, does this make the behaviour difference when autocompleting text that's not following a `.`? (E.g. with the example in the original post.) Just want to check that autocomplete for property accesses isn't special-cased at all. Thanks!

**Comment by jrieken** [comment]
No, unfortunately not. Due to #43429 a non whitespace prefix is needed...

**Comment by jrieken** [comment]
I am open for suggestions, if you have a better idea to fix #43429 I can make such changes

**Comment by varkor** [comment]
@jrieken: if I understand correctly, the main problem with #43429 is that VSCode doesn't/can't take the context into account (e.g. inside an object literal, etc.) when providing suggestions, so the special-casing for empty lines is to give more helpful behaviour in languages where these contexts are common?

It seems the simplest way to fix this would be to add another setting to enable/disable this (which could be done on a per-language setting) — for example, it clearly doesn't make sense for plaintext files (which is what I'm concerned with for this issue). It seems counterintuitive for `recentlyUsed` to be special-cased on empty lines (based on its description). Maybe even another option `recentlyUsedByContext` (which could just have the empty line special-casing for now, but which would have the possibility of being upgraded to a more semantic approach in the future) would be a solution?

**Comment by jrieken** [comment]
> @jrieken: if I understand correctly, the main problem with #43429 is that VSCode doesn't/can't take the context into account (e.g. inside an object literal, etc.) 

Yes, there is no context indeed and there is currently no way to feed context into IntelliSense. I believe the problem with #43429 is that snippets show even though member completion is active (which we don't know) and snippets don't really make sense there.

**Comment by varkor** [comment]
@jrieken: this may be a separate issue entirely, but does the autocomplete take into account what type of suggestion was selected previously? I have a file in which `test` is suggested as both a name and a snippet, but regardless of which I selected previously, the name is always suggested first.
